John password

jachrostek opened the pull request:
Added the ability for users to have a password. When adding a new user, a password must  be supplied. When signing in, a password must also be supplied.

jachrostek committed:
Added Design Document

jachrostek committed:
Added Design Document

jachrostek committed:
Adding Password Implementation

jachrostek committed:
Adding Password Implementation

jachrostek committed:
Merge pull request #1 from jachrostek/password

Adding Password Implementation

jachrostek committed:
Merge branch 'master' of https://github.com/google/codeu_project_2017 into password

jachrostek committed:
Merge pull request #2 from jachrostek/password

Password

jachrostek committed:
Conversations can now store a message

jachrostek committed:
Conversations can now store a message

jachrostek committed:
Merge branch 'master' of https://github.com/jachrostek/codeu_project_2017 into privateConversation

jachrostek committed:
Merge pull request #3 from jachrostek/privateConversation

Private conversation

jachrostek committed:
changes back to original run client

jachrostek committed:
Modified internal working of password implementaion, added Sign-In with password to the GUI.

jachrostek committed:
Merge branch 'master' into JohnPassword

jachrostek committed:
Merge pull request #4 from jachrostek/JohnPassword

John password

jachrostek committed:
Minor changes

jachrostek committed:
Merge branch 'JohnPassword' of https://github.com/jachrostek/codeu_project_2017 into JohnPassword

jachrostek committed:
Merge pull request #5 from jachrostek/JohnPassword

John password

jachrostek committed:
Minor modifications to class roles for implementing the password feature

jachrostek committed:
Merge pull request #6 from jachrostek/PasswordUpdate

Minor modifications to class roles for implementing the password feature

jachrostek committed:
Update ClientConversation.java

jachrostek committed:
Update Chat.java

JacobWarner commented:
Overall, it looks pretty solid to me. Although, what was the reasoning behind removing Hub.java? Also, could you revise your comment for the PR? It seems there is a lot more work than just on password implementation (i.e. relay server). I'd like to know why you made the changes you did, as it's a little overwhelming going through 28 changed files and deciphering it myself. Thank you!

Commit:
Merge branch 'develop' into JohnPassword

jachrostek committed:
Update Server.java

jachrostek commented:
I don't think I changed Hub.java and I don't see it being removed when I go through the files changed section. If I did remove it, it was a mistake. I agree that more detail would be better for the comment and I will try to do a better job of adding more information in the future. 

I added a Password class whose primary purpose is to return a passwordHash string from a password string and a salt string. The salt is a random string generated through a static method of the Password class. The Password class uses a cryptographic hashing scheme to return a passwordHash. Cryptographic hashing was chosen to store the passwords because it provides a good balance of simplicity and security. It is able to create the passwordHash from the password but it can't generate the password from the passwordHash. This is idea because the actual password value is not stored. The purpose of the salt is to add more variety and therefore spread out the hashed values. Salt also guarantees that if different users have the same password, they hash to different values. Consequently, it is more difficult for someone to try and determine the password values even if they have access to the passwordHash values. 

The variables PasswordHash and Salt were added to the User class along with methods to return those values. The Chat and SimpleGUI were changed such that users must add a password between 6 and 14 characters long in order to create a new user. Additionally, these classes ask for a password when signing in. There were several methods throughout the program that were involved with creating a new user. These method were changed to accommodate the PasswordHash and Salt. The main changes were adding parameters to the method signature and either passing on the values to another method call or serializing them so they could be sent back and forth between the client and server. The ClientUser class  determines whether or not the password is correct when signing in. It checks to see if the password attempt hashes to the passwordHash created with the original password. 

Hopefully that clears things up. I could also provide more detail if needed.

JacobWarner commented:
I'd say that clears it up! Thank you for the details. It makes much more sense to me now. I'm gonna go ahead and merge it into develop, but we'll wait until Travis and Kevin are able to look through develop before merging everything into master.

Good job!
